File: src/infrastructure/gemini/adapter.py
import logging
import asyncio
import base64
import os
import json
from typing import AsyncGenerator, Optional, Any, Union, List, Callable, Dict
from .rest_client import GeminiRestClient
from . import errors
from .retry_handler import RetryHandler
from .snapshot_manager import SnapshotManager
from src.infrastructure.gemini.utils import estimate_tokens

logger = logging.getLogger(__name__)

class GeminiAdapter:
    def __init__(self):
        self.client = GeminiRestClient()
        self.history: List[dict] = []
        self._current_model: Optional[str] = None
        self.system_instruction: Optional[str] = None
        self.temperature: float = 0.7
        self.retry_handler = RetryHandler()
        self.snapshot_manager = SnapshotManager()

    def set_retry_callback(self, callback: Callable[[str], None]):
        self.retry_handler.on_retry = callback

    def start_chat(self, model_name: str, system_instruction: str, temperature: float):
        self._current_model = model_name
        self.system_instruction = system_instruction
        self.temperature = temperature
        self.history = []

    def _build_payload(self) -> Dict[str, Any]:
        contents = list(self.history)
        payload = {"contents": contents, "generationConfig": {"temperature": self.temperature}}

        if self.system_instruction:
            if self._current_model and "gemma-3" in self._current_model.lower():
                if contents:
                    first_msg = contents[0]
                    if first_msg["role"] == "user":
                        first_msg["parts"][0]["text"] = f"{self.system_instruction}\n\n{first_msg['parts'][0]['text']}"
                else:
                    payload["contents"].append({"role": "user", "parts": [{"text": self.system_instruction}]})
            else:
                payload["systemInstruction"] = {"parts": [{"text": self.system_instruction}]}
        
        return payload

    async def send_message_stream(self, message: str, role: str = "user") -> AsyncGenerator[str, None]:
        parts = [{"text": message}]
        self.history.append({"role": role, "parts": parts})
        
        payload = self._build_payload()
        full_text = ""
        try:
            # Await the async generator from the client
            async for chunk in await self.client.generate_content(self._current_model or "", payload, stream=True):
                full_text += chunk
                yield chunk
        except Exception as e:
            logger.error(f'Streaming error: {e}')
            yield f'Error: {str(e)}'
        
        self.history.append({"role": "model", "parts": [{"text": full_text}]})

    async def send_message(self, message: Union[str, bytes], role: str = "user") -> Any:
        parts = [{"text": message}] if isinstance(message, str) else [{"inline_data": {"mime_type": "audio/wav", "data": base64.b64encode(message).decode()}}]
        self.history.append({"role": role, "parts": parts})
        
        payload = self._build_payload()

        try:
            # Await the async retry handler
            response = await self.retry_handler.execute(
                self.client.generate_content, 
                self._current_model or 'gemini-2.0-flash', 
                payload
            )
            
            if not response:
                logger.error('API returned empty response')
                return 'Error: API returned empty response'

            candidates = response.get('candidates', [])
            if not candidates:
                logger.warning(f'No candidates returned. Response: {response}')
                return 'Error: No response candidates returned (possible safety block).'

            candidate = candidates[0]
            finish_reason = candidate.get('finishReason')
            
            if finish_reason == 'SAFETY':
                logger.warning(f'Response blocked by safety filters. Response: {response}')
                return 'Error: Response blocked by safety filters.'

            content = candidate.get('content', {})
            parts = content.get('parts', [])
            if not parts:
                logger.warning(f'No parts in content. Response: {response}')
                return 'Error: No content parts returned.'

            text_content = "".join([p.get('text', '') for p in parts])
            if text_content:
                self.history.append({"role": "model", "parts": [{"text": text_content}]})
                return text_content
            return 'Error: Empty response'

        except Exception as e:
            logger.exception(f'Critical error in send_message: {e}')
            return f'Error: {str(e)}'

    def inject_message(self, role: str, content: str):
        self.history.append({"role": role, "parts": [{"text": content}]})

    def load_history(self, history_data: List[dict], model_name: str, temperature: float):
        self._current_model = model_name
        self.temperature = temperature
        self.history = history_data

    async def get_available_models(self) -> List[str]:
        try:
            # Await the async client call
            models = await self.client.list_models()
            return [m['name'] for m in models if 'gemini' in m['name'] or 'gemma' in m['name']]
        except Exception as e:
            logger.error(f"Error fetching models: {e}")
            return ['gemini-2.0-flash', 'gemini-1.5-flash']

    async def check_availability(self, model_name: str) -> bool:
        return model_name in await self.get_available_models()

    def hard_reset(self, history_override: Optional[List[dict]] = None):
        self.history = history_override if history_override is not None else []

    def pop_last_message(self):
        if self.history:
            self.history.pop()

    def save_snapshot(self) -> bool:
        last_msg = str(self.history[-1]) if self.history else ""
        return self.snapshot_manager.save(self.system_instruction, self.history, last_msg)

    def get_history(self) -> List[dict]:
        return self.history

    def prune_history(self, max_tokens: int):
        """Prunes the history to stay within the token limit, preserving the most recent context."""
        from src.infrastructure.gemini.utils import estimate_tokens
        if len(self.history) <= 2:
            return

        while len(self.history) > 2 and estimate_tokens(self.history) > max_tokens:
            self.history.pop(0)

    def count_tokens(self, history: List[dict]) -> int:
        return estimate_tokens(history)

File: src/infrastructure/gemini/adapter.js
import axios from "axios";
import logger from "../../core/logger.js";
import BaseAdapter from "../baseAdapter.js";
import IRCFormatter from "../../utils/ircFormatter.js";
import RetryHandler from "../../utils/retryHandler.js";
import {
  GeminiSafetyError,
  GeminiEmptyResponseError,
} from "../../core/exceptions.js";

/**
 * Adapter for the Google Gemini API, handling communication, rate limiting, and history.
 * @augments BaseAdapter
 */
class GeminiAdapter extends BaseAdapter {
  /**
   * Initializes the GeminiAdapter with necessary dependencies and configurations.
   * @param {object} ui - The UI interface for displaying information.
   * @param {object} configManager - The configuration manager.
   * @param {string} [userNickname] - The nickname of the user.
   * @param {string} [agentNickname] - The nickname of the agent.
   */
  constructor(
    ui,
    configManager,
    userNickname = "user",
    agentNickname = "assistant",
  ) {
    super(ui, configManager, userNickname, agentNickname);
    this.apiKey = process.env.GEMINI_API_KEY || process.env.GOOGLE_API_KEY;
    this.history = [];
    this.currentModel = "models/gemma-4-26b-a4b-it";
    this.systemInstruction = null;
    this.temperature = 0.7;
    this.lastPayload = null;
    this.renderingMode = "IRC";
    this.lastRequestTime = 0;

    this.client = axios.create({
      timeout: 600000,
    });
    this.retryHandler = new RetryHandler();
    this.recoverableErrors = [
      "Empty response from Gemini",
      "No response candidates returned",
      "No content parts returned",
      "Response blocked by safety filters.",
    ];
  }

  /**
   * Sets the rendering mode for the adapter.
   * @param {string} mode - The rendering mode to use.
   */
  setRenderingMode(mode) {
    this.renderingMode = mode;
  }

  /**
   * Configures the chat session with a specific model, system instruction, and temperature.
   * @param {string} modelName - The name of the model to use.
   * @param {string} systemInstruction - The system instruction for the model.
   * @param {number} [temperature] - The temperature for generation.
   */
  startChat(modelName, systemInstruction, temperature = 0.7) {
    this.currentModel = modelName || this.currentModel;
    this.systemInstruction = systemInstruction;
    this.temperature = temperature;
  }

  /**
   * Builds the final payload for the Gemini API request based on conversation history.
   * @private
   * @returns {object} The constructed payload.
   */
  _buildPayload() {
    const contents = this.history.map((entry) => {
      const role = entry.role === "server" ? "user" : entry.role;

      const parts = entry.parts.map((part) => {
        if (part && typeof part === "object" && part.inline_data) {
          return part;
        }

        let textContent;
        if (typeof part === "object" && part.text) {
          textContent = part.text;
        } else if (typeof part === "string") {
          textContent = part;
        } else {
          textContent = JSON.stringify(part);
        }

        const formattedText = this.formatTextForPayload(
          role,
          textContent,
          entry.timestamp,
        );

        return {
          text: formattedText,
        };
      });

      return { role, parts };
    });

    const payload = {
      contents,
      generationConfig: {
        temperature: this.temperature,
        top_p: 0.95,
        top_k: 64,
      },
    };

    if (this.systemInstruction) {
      payload.systemInstruction = {
        parts: [{ text: this.systemInstruction }],
      };
    }

    return payload;
  }

  /**
   * Removes thought/reasoning tags and timestamps from the model response.
   * @private
   * @param {string} text - The raw text to filter.
   * @returns {string} The cleaned text.
   */
  _filterThoughts(text) {
    if (!text) return "";
    let cleaned = text.replace(/<(thought|reasoning)>[\s\S]*?<\/\1>/gi, "");
    cleaned = cleaned.replace(/^(\[\d{2}:\d{2}:\d{2}\]\s*<[^>]+>\s*)+/g, "");
    return cleaned.trim();
  }

  /**
   * Converts various content types into Gemini API compatible parts.
   * @private
   * @param {string|Array|object} content - The content to convert.
   * @returns {Array<object>} An array of Gemini API parts.
   */
  _createParts(content) {
    if (Array.isArray(content)) {
      return content.flatMap((item) => {
        if (typeof item === "string") return [{ text: item }];
        if (item && typeof item === "object" && item.mime_type && item.data) {
          return [
            { inline_data: { mime_type: item.mime_type, data: item.data } },
            { text: `Image resolution: ${item.resolution || "unknown"}` },
          ];
        }
        return [{ text: item === undefined ? "" : JSON.stringify(item) }];
      });
    }
    if (typeof content === "string") {
      return [{ text: content }];
    }
    if (
      content &&
      typeof content === "object" &&
      content.mime_type &&
      content.data
    ) {
      return [
        { inline_data: { mime_type: content.mime_type, data: content.data } },
        { text: `Image resolution: ${content.resolution || "unknown"}` },
      ];
    }
    return [{ text: content === undefined ? "" : JSON.stringify(content) }];
  }

  /**
   * Records a message in the conversation history.
   * @private
   * @param {string} role - The role of the message sender.
   * @param {string|Array|object} content - The content of the message.
   * @param {string|null} [timestamp] - The timestamp of the message.
   */
  _recordMessage(role, content, timestamp = null) {
    const ts = timestamp || IRCFormatter.getTimestamp();
    const parts = this._createParts(content);

    this.history.push({
      role,
      parts,
      timestamp: ts,
    });
  }

  /**
   * Sends a message to the Gemini API and returns the response.
   * @param {string|Array|object} message - The message to send.
   * @param {string} [role] - The role of the sender.
   * @returns {Promise<string>} The processed response text.
   * @throws {Error} If the API request fails.
   */
  async sendMessage(message, role = "user") {
    await this._applyRateLimit();
    this._recordMessage(role, message);
    this._enforceContextLimit(); // Apply strict context boundary before API call
    const historyLengthBefore = this.history.length;

    /**
     * Executes the API request with retry logic.
     * @returns {Promise<string>} The processed response text.
     */
    try {
      return await this.retryHandler.execute(async () => {
        try {
          const payload = this._buildPayload();
          this.lastPayload = payload;

          const modelName = this.currentModel.replace(/^models\//, "");
          const url = `https://generativelanguage.googleapis.com/v1beta/models/${modelName}:generateContent?key=${this.apiKey}`;

          const response = await this.client.post(url, payload);
          return this._processApiResponse(response.data);
        } catch (error) {
          throw this._wrapApiError(error);
        }
      }, {
      recoverableErrors: this.recoverableErrors,
      /**
       * @param {number} attempt - The current attempt number.
       * @param {Error} error - The error that triggered the retry.
       * @param {string} formattedDelay - The formatted delay string.
       */
      onRetry: (attempt, error, formattedDelay) => {
        logger.warn(`[GeminiAdapter] Retrying in ${formattedDelay}... (${attempt}/15) due to: ${error.message}`);
        if (this.ui && this.ui.displayInfo) {
          this.ui.displayInfo(`Reintentando Gemini en ${formattedDelay}... (${attempt}/15) | Error: ${error.message}`);
        }
      }
    });
    } catch (error) {
      if (this.history.length === historyLengthBefore) {
        this.popLastMessage();
      }
      throw error;
    }
  }

  /**
   * Processes the raw API response and extracts the text content.
   * @private
   * @param {object} data - The raw response data from the API.
   * @returns {string} The filtered response text.
   * @throws {GeminiEmptyResponseError|GeminiSafetyError} If the response is invalid or blocked.
   */
  _processApiResponse(data) {
    const { candidates } = data;
    if (!candidates || candidates.length === 0) {
      throw new GeminiEmptyResponseError(
        "No response candidates returned (possible safety block).",
      );
    }

    const candidate = candidates[0];
    if (candidate.finishReason === "SAFETY") {
      throw new GeminiSafetyError("Response blocked by safety filters.");
    }

    const { content } = candidate;
    if (!content || !content.parts || content.parts.length === 0) {
      throw new GeminiEmptyResponseError("No content parts returned.");
    }

    let textContent = content.parts.map((p) => p.text).join("");
    textContent = this._filterThoughts(textContent);

    if (textContent) {
      this._recordMessage("model", textContent);
      return textContent;
    }

    throw new GeminiEmptyResponseError("Empty response");
  }

  /**
   * Wraps API errors into a standardized APIError format.
   * @private
   * @param {Error} error - The original error.
   * @returns {Error} The wrapped APIError.
   */
  _wrapApiError(error) {
    const errorMsg = error.response?.data?.error?.message || error.message;
    const wrappedError = new Error(errorMsg);
    wrappedError.name = "APIError";
    wrappedError.response = error.response;
    wrappedError.code = error.code;
    return wrappedError;
  }

  /**
   * Manually injects a message into the conversation history.
   * @param {string} role - The role of the message sender.
   * @param {string|Array|object} content - The content of the message.
   * @param {string|null} [timestamp] - The timestamp of the message.
   */
  injectMessage(role, content, timestamp = null) {
    this._recordMessage(role, content, timestamp);
  }

  /**
   * Resets the conversation history.
   * @param {Array|null} [historyOverride] - Optional new history to initialize with.
   */
  hardReset(historyOverride = null) {
    this.history = historyOverride || [];
  }

  /**
   * Retrieves the current conversation history.
   * @returns {Array} The conversation history.
   */
  getHistory() {
    return this.history;
  }

  /**
   * Removes the last message from the conversation history.
   */
  popLastMessage() {
    if (this.history.length > 0) {
      this.history.pop();
    }
  }

  /**
   * Estimates the token count based on character length heuristic.
   * Accounts for the 'parts' array structure specific to the Gemini API.
   * @param {string} systemInstruction - The system instruction text.
   * @param {Array} history - The conversation history.
   * @returns {number} The estimated token count.
   */
  countTokens(systemInstruction, history) {
    const CHARS_PER_TOKEN = 3.5;
    const systemChars = systemInstruction?.length || 0;
    const historyChars = history.reduce((acc, msg) => {
      const partsLength = (msg.parts || []).reduce((partAcc, part) => {
        if (typeof part.text === "string") {
          return partAcc + part.text.length;
        }
        return partAcc;
      }, 0);
      return acc + partsLength;
    }, 0);
    return Math.ceil((systemChars + historyChars) / CHARS_PER_TOKEN);
  }

  /**
   * Fetches the list of available models from the Gemini API.
   * @returns {Promise<string[]>} A list of available model names.
   */
  async getAvailableModels() {
    try {
      const url = `https://generativelanguage.googleapis.com/v1beta/models?key=${this.apiKey}`;
      const response = await this.client.get(url);
      const { data } = response;
      return data.models.map((m) => m.name);
    } catch (error) {
      logger.error(`Error fetching models: ${error.message}`);
      return ["models/gemini-2.0-flash", "models/gemini-1.5-flash"];
    }
  }

  /**
   * Checks if a specific model is available and responsive.
   * @param {string} modelName - The name of the model to check.
   * @returns {Promise<boolean>} True if available, false otherwise.
   */
  async checkAvailability(modelName) {
    try {
      const name = modelName.replace(/^models\//, "");
      const url = `https://generativelanguage.googleapis.com/v1beta/models/${name}:generateContent?key=${this.apiKey}`;
      const payload = { contents: [{ role: "user", parts: [{ text: "hi" }] }] };
      await this.client.post(url, payload);
      return true;
    } catch (error) {
      const status = error.response?.status;
      if (status === 404 || status === 400) return false;
      return true;
    }
  }
}

export default GeminiAdapter;
